pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -571,6 +596,145 @@ func (a Adapter) generateDeploymentObjectMeta(deployment *appsv1.Deployment, lab
+			if remoteK.GroupVersionKind() == kclient.SecretGVK {
+				var remoteSecretHasLocalServiceBindingOwner bool
+				if isSBOSupported {
+					// if SBO is installed, check if its owner reference is a servicebinding present in the devfile;
+					// if the owner matches, consider it a match found
+					remoteSecretHasLocalServiceBindingOwner = func() bool {
+						for _, owner := range remoteK.GetOwnerReferences() {
+							if owner.Kind == kclient.ServiceBindingKind && owner.Name == devfileK.GetName() {
+								return true
+							}
+						}
+						return false
+					}()
+
+				} else {
+					// if SBO is not installed, check if it is a servicebinding secret
+					remoteSecretHasLocalServiceBindingOwner = service.IsLinkSecret(remoteK.GetLabels()) &&
+						remoteK.GetLabels()[service.LinkLabel] == devfileK.GetName() &&
+						remoteK.GetNamespace() == currentNamespace
+				}
+				if remoteSecretHasLocalServiceBindingOwner {
+					matchFound = true
+					break
+				}
+			}
+		}
+		if !matchFound {
+			// only remove servicebinding secrets if the SBO is not installed
+			if !isSBOSupported && remoteK.GroupVersionKind() == kclient.SecretGVK && service.IsLinkSecret(remoteK.GetLabels()) {
+				serviceBindingSecretsToRemove = append(serviceBindingSecretsToRemove, remoteK)
+			} else {
+				objectsToRemove = append(objectsToRemove, remoteK)
+			}
+		}
+	}
+	return objectsToRemove, serviceBindingSecretsToRemove, nil
+}
+
+// deleteRemoteResources takes a list of remote resources to be deleted
+func (a Adapter) deleteRemoteResources(objectsToRemove []unstructured.Unstructured) error {
+	// Delete the resources present on the cluster but not in the Devfile
+	for _, objectToRemove := range objectsToRemove {
+		spinner := log.Spinnerf("Deleting Kubernetes resource: %s/%s", objectToRemove.GetKind(), objectToRemove.GetName())
+
+		gvr, err := a.kubeClient.GetGVRFromGVK(objectToRemove.GroupVersionKind())
+		if err != nil {
+			return fmt.Errorf("unable to get information about Kubernetes resource: %s/%s: %s", objectToRemove.GetKind(), objectToRemove.GetName(), err.Error())
+		}
+
+		err = a.kubeClient.DeleteDynamicResource(objectToRemove.GetName(), gvr, true)
+		if err != nil {
+			if !kerrors.IsNotFound(err) {
+				spinner.End(false)
+				return fmt.Errorf("unable to delete Kubernetes resource: %s/%s: %s", objectToRemove.GetKind(), objectToRemove.GetName(), err.Error())
+			}
+
+			klog.V(4).Infof("Failed to delete Kubernetes resource: %s/%s; resource not found", objectToRemove.GetKind(), objectToRemove.GetName())
+		}
+		spinner.End(true)
+	}
+	return nil
+}
+
+// deleteServiceBindingSecret takes a list of Service Binding secrets that should be deleted;
+// this is mainly helpful when Service Binding Operator is not installed on the cluster
+func (a Adapter) deleteServiceBindingSecret(serviceBindingSecretsToRemove []unstructured.Unstructured, deployment *appsv1.Deployment) error {
+	for _, secretToRemove := range serviceBindingSecretsToRemove {
```

Comment:
Same point about sending out requests in parallel.

Reply:
I am not sure doing that here is a good idea. There are possible independent secrets(only one secret most of the time), but only one deployment that is linked to all the secrets. When the unbinding happens, it modifies the deployment by removing any reference to the secret, and then proceeds to delete the secret.

Reply:
Alright, fine then to keep it as is here.